[JIT] improve mayContainAlias

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31841 [JIT] implement tuple constants
* #31840 [JIT] remove tuple logic in constant propagation
* **#31839 [JIT] improve mayContainAlias**
* #31791 Add str[] float[] constants resubmit

There are a number of improvements that can be made to `mayContainAlias`, which I would like to do in follow ups. For now, this is an easy one.

Differential Revision: [D19439516](https://our.internmc.facebook.com/intern/diff/D19439516)

Commit:
improve mayContainAlias

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 24a738c9:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=9e9bfbfd8d893cdf8205db67f220551214787d7f&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 9e9bfbfd since Jan 02
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  Since your merge base is older than `viable/strict`, run these commands:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=24a738c986363ee29feb9fff4290cf42721304cc).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_macos_10_13_py3_test](https://dr.pytorch.org/build-details.html?build_id=79943179) from Jan 02 until Jan 02



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

suo commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -853,28 +853,41 @@ bool AliasDb::mayContainAlias(
   for (const auto& val : a) {
     if (cannotCheckAliasContainment(val)) {
-      return true;
+      a_cannot_check_containment = true;
```
nit: I think it might be a bit clearer if we just do `return b_elements.size() == 0` with a comment here instead of keeping a variable alive.

eellison replied:
that doesn't cover the case where b contains immutable elements

eellison replied:
`b_elements.size() == 0` will always be true because we haven't iterated over `b` yet

suo replied:
mmm makes sense lol

Commit:
Update on "[JIT] improve mayContainAlias"


There are a number of improvements that can be made to `mayContainAlias`, which I would like to do in follow ups. For now, this is an easy one. 


[ghstack-poisoned]

Commit:
Update on "[JIT] improve mayContainAlias"


There are a number of improvements that can be made to `mayContainAlias`, which I would like to do in follow ups. For now, this is an easy one. 


[ghstack-poisoned]

Commit:
Update on "[JIT] improve mayContainAlias"


There are a number of improvements that can be made to `mayContainAlias`, which I would like to do in follow ups. For now, this is an easy one. 


[ghstack-poisoned]

Commit:
Update on "[JIT] improve mayContainAlias"


There are a number of improvements that can be made to `mayContainAlias`, which I would like to do in follow ups. For now, this is an easy one.

Differential Revision: [D19439516](https://our.internmc.facebook.com/intern/diff/D19439516)

[ghstack-poisoned]
